src/libs/protoboard/blackboard_manager.h

```diff
@@ -0,0 +1,331 @@
+{
+public:
+	/** Main thread constructor
+	 * @param msg_handler A pointer to the thread that receives incoming ProtoBuf messages */
+	BlackboardManager(ProtobufThead *msg_handler);
+
+	/** Helper for other classes to get access to the blackboard
+	 * @return Pointer to the blackboard used by this thread */
+	fawkes::BlackBoard *get_blackboard();
+
+	/** The ProtoBuf sender must be initialized after construction to beak a dependency loop
+	 * @param sender The initialized ProtobufSender */
+	void set_protobuf_sender(AbstractProtobufSender *sender);
+
+protected:
+	virtual void init() override;
+	virtual void finalize() override;
+	virtual void loop() override;
+
+	/** Act on a given message on a given blackboard interface. Must be implemented by the user.
+	 * @tparam the blackboard interface type
+	 * @tparam the blackboard message type
+	 * @param iface a pointer to the concrete interface
+	 * @param msg a pointer to the concrete message that came in on that interface */
+	template <class InterfaceT, class MessageT>
+	void handle_message(InterfaceT *iface, MessageT *msg);
+
+private:
+	friend AbstractProtobufSender;
+
+	ProtobufThead *                         message_handler_;
+	fawkes::ProtobufPeerInterface *         peer_iface_;
+	pb_conversion_map                       bb_receiving_interfaces_;
+	fawkes::BlackBoardOnMessageWaker *      on_message_waker_;
+	unsigned int                            next_peer_idx_;
+	std::unique_ptr<AbstractProtobufSender> pb_sender_;
+
+	void add_peer(fawkes::ProtobufPeerInterface *iface, long peer_id);
+
+	template <class MessageT, class InterfaceT>
+	bool handle_message_type(InterfaceT *iface);
+
+	template <class InterfaceT>
+	struct on_interface
+	{
+		InterfaceT *       iface;
+		BlackboardManager *manager;
+
+		on_interface(InterfaceT *iface, BlackboardManager *manager) : iface(iface), manager(manager)
+		{
+		}
+
+		template <class MessageT>
+		bool
+		handle_msg_types()
+		{
+			return manager->handle_message_type<MessageT>(iface);
+		}
+
+		// This template is disabled if MessageTs is {} to resolve ambiguity
+		template <class MessageT1, class... MessageTs>
+		typename std::enable_if<(sizeof...(MessageTs) > 0), bool>::type
+		handle_msg_types()
+		{
+			return handle_msg_types<MessageTs...>() || handle_msg_types<MessageT1>();
+		}
+	};
+};
+
+template <class... IfaceManagerTs>
+ProtobufSender<IfaceManagerTs...>::ProtobufSender(BlackboardManager *bb_mgr)
+: AbstractProtobufSender(bb_mgr)
+{
+}
+
+template <class... IfaceManagerTs>
+void
+ProtobufSender<IfaceManagerTs...>::init()
+{
+	boost::fusion::for_each(bb_sending_interfaces_, [this](auto &iface_mgr) {
+		iface_mgr.init(this->bb_manager->get_blackboard(), this->bb_manager);
+	});
+}
+
+template <class... IfaceManagerTs>
+void
+ProtobufSender<IfaceManagerTs...>::finalize()
+{
+	boost::fusion::for_each(bb_sending_interfaces_,
+	                        [this](auto &iface_mgr) { iface_mgr.finalize(); });
```

Comment:
See above, I'd say this can go into the destructor

Reply:
Yeah, but the initialization actually cannot happen in the constructor, so for congruence's sake I made it `init/finalize`.